### PR TITLE
add flag dnn.enabled

### PR DESCRIPTION
**Pull request by nouiz**
This allow to always disable cudnn without corner case where it is partially installed.

fix gh-3962

### CODE REVIEW

**Comment by nouiz** [comment]
I did the change.

File: doc/library/sandbox/cuda/dnn.txt [hunk 1]
```diff
@@ -38,8 +38,10 @@ By default, Theano will detect if it can use cuDNN. If so, it will use
 it.  If not, Theano optimizations will not introduce cuDNN ops. So
 Theano will still work if the user did not introduce them manually.
 
-To get an error if Theano can not use cuDNN, use this Theano flag:
-``optimizer_including=cudnn``.
+The recently added Theano flag :attr:`allow_gc <config.dnn.enabled>`
```
Comment (lamblin) [inline doc/library/sandbox/cuda/dnn.txt]: "allow_gc"?

File: doc/library/config.txt [hunk 2]
```diff
@@ -567,6 +567,19 @@ import theano and print the config variable, as in:
...
 
+.. attribute:: dnn.enabled
+
+  String value: ``auto``, ``True``, ``False``
+
+  Default ``auto``
+
+  If ``auto``, auto detect if cudnn is there, if not, raise an error.
+
+  If ``True``, if we can't use cudnn, raise an error.
```
Comment (lamblin) [inline doc/library/config.txt]: I don't understand the difference between "auto" and "True", then.

Comment (nouiz) [inline doc/library/config.txt]: auto, autodetect. If present use it, if not present, don't use it, but don't raise an error.

True, raise an error if no present. I'll update the doc.

**Comment by nouiz** [comment]
I updated the PR.

File: doc/library/config.txt [hunk 3]
```diff
@@ -567,6 +567,21 @@ import theano and print the config variable, as in:
...
+
+  Default ``auto``
+
+  If ``auto``, auto detect if cudnn is there, if not, do not use it,
+  but do not raise an error.
+
+  If ``True``, if we can't use cudnn, raise an error.
+
+  If ``False``, do not test if cudnn is there and don't use it.
+  Useful to disable it completly when the auto detection crash in some
```
Comment (lamblin) [inline doc/library/config.txt]: "completely", "crashes". I would remove "in some strange case"

File: doc/library/sandbox/cuda/dnn.txt [hunk 4]
```diff
@@ -38,8 +38,10 @@ By default, Theano will detect if it can use cuDNN. If so, it will use
 it.  If not, Theano optimizations will not introduce cuDNN ops. So
 Theano will still work if the user did not introduce them manually.
 
-To get an error if Theano can not use cuDNN, use this Theano flag:
-``optimizer_including=cudnn``.
+The recently added Theano flag :attr:`dnn.enabled <config.dnn.enabled>`
+allow to change the default behavior to force it or disable it. You
```
Comment (lamblin) [inline doc/library/sandbox/cuda/dnn.txt]: "allows"

File: doc/library/sandbox/cuda/dnn.txt [hunk 5]
```diff
@@ -38,8 +38,10 @@ By default, Theano will detect if it can use cuDNN. If so, it will use
...
+can also get an error with older Theano version if cuDNN can not be
+used, by using this Theano flag: ``optimizer_including=cudnn``.
```
Comment (lamblin) [inline doc/library/sandbox/cuda/dnn.txt]: I don't understand this sentence.

File: theano/configdefaults.py [hunk 6]
```diff
@@ -329,6 +329,11 @@ def f(suffix=suffix):
              "Location of the cudnn header (defaults to the cuda root)",
              StrParam(default_dnn_path('lib64')))
 
+AddConfigVar('dnn.enabled',
+             "If False, disable cudnn",
+             StrParam("auto", "True", "False"),
```
Comment (lamblin) [inline theano/configdefaults.py]: There should be an explanation here about what "True" and "auto" mean, as well.

**Comment by nouiz** [comment]
I did the fix and added some other doc update related to dnn in the last commit.

**Comment by lamblin** [comment]
OK for me when Travis passes.

**Pull request merged**